### PR TITLE
Make LRU prefix caching eviction policy only evict child blocks

### PR DESCRIPTION
**Pull request by santhnm2**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
The LRU prefix eviction policy only considered timestamps when selecting block to evict, but this could lead to parent blocks being evicted and leaving holes in the prefix chain. Instead, this PR ensures that parent <-> child relationships are taken into account by the LRU eviction policy so that only leaf node blocks are evicted. This PR also makes ref-count updates happen BEFORE allocation rather than AFTER allocation so that pending matched blocks are not accidentally evicted because their ref-counts were 0.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/contexts/dynamic_context.py
Comment (lmcafee-nvidia) [inline megatron/core/inference/contexts/dynamic_context.py:1]: In [`add_request()`](https://github.com/NVIDIA/Megatron-LM/blob/3c08be510fb8653921215a35f2e60736c7827564/megatron/core/inference/contexts/dynamic_context.py#L2948-L2978), `matched_block_ids` are selected before `allocate_memory_blocks()` runs, but their refcounts are incremented only after allocation. If allocation triggers LRU eviction, those pending matches still have `ref_count == 0` and remain eviction candidates.

For example, assume three full KV slots:

- Cached chain `H0/S0 → H1/S1`, with timestamp `1`.
- Unrelated cached root `HX/SX`, with timestamp `10`.
- Incoming prompt `H0 → H1 → H2`.

Here `H1` is a leaf, `HX` is a separate root, and `H2` is not cached yet. Matching returns `[S0, S1]` and the request needs one new block. Because `S0` and `S1` are not pinned yet, descendant-first LRU evicts the older leaf `H1/S1`. Allocation immediately reuses `S1`, while the stale match list still contains it, producing:

`block_table = [S0, S1, S1]`

Registering `H2` then overwrites `S1` and leaves `H2 → missing H1`, corrupting the request’s KV history and recreating the dangling-parent state this PR is intended to prevent. The correct result is to protect `S0/S1`, evict `SX`, and produce `[S0, S1, SX]`.

Could we protect matched IDs during both availability checking and eviction/allocation—either by passing an exclusion set or by pinning them with rollback on failure? Please also add a full-cache partial-hit `add_request()` regression that verifies distinct block IDs and a contiguous `H0 → H1 → H2` hash chain.

The allocate-before-pin ordering predates this PR, but it directly invalidates the invariant being introduced here, so I consider it a merge blocker.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 1]
```diff
@@ -2875,14 +2875,20 @@ def check_availability(self, req: DynamicInferenceRequest) -> Tuple[bool, bool,
...
             self.active_token_count + effective_prefill_chunk_length <= self.max_tokens
         )
-        kv_cache_available = self.kv_block_allocator.is_memory_available(num_blocks_from_pool)
+        # Matched blocks are currently evictable (ref_count == 0) but add_request
+        # will pin them before allocating, so they cannot serve as free capacity
+        # for the new blocks this request needs. Exclude them from the evictable
+        # count so availability matches what allocation can actually satisfy.
+        kv_cache_available = self.kv_block_allocator.is_memory_available(
+            num_blocks_from_pool, reserved_evictable=len(matched_block_ids)
+        )
```
[the remark this answers is not in the document]
Comment (santhnm2) [reply megatron/core/inference/contexts/dynamic_context.py]: Addressed

File: megatron/core/inference/contexts/kv_block_allocator.py [hunk 2]
```diff
@@ -352,13 +405,53 @@ def evict_lru_blocks(self, num_blocks_needed: int) -> bool:
...
+        depth = torch.zeros(num_cached, dtype=torch.int64)
+        for _ in range(num_cached):
+            new_subtree_max = subtree_max.clone()
+            new_subtree_max.scatter_reduce_(
+                0, parent_idx[has_parent], subtree_max[has_parent], reduce="amax"
+            )
+            new_depth = torch.where(has_parent, depth[parent_safe] + 1, depth)
+            if torch.equal(new_subtree_max, subtree_max) and torch.equal(new_depth, depth):
+                break
+            subtree_max, depth = new_subtree_max, new_depth
```
[the remark this answers is not in the document]
Comment (santhnm2) [reply megatron/core/inference/contexts/kv_block_allocator.py]: Added a TODO to revisit alternative solutions if perf is an issue

File: megatron/core/inference/contexts/dynamic_context.py
Comment (santhnm2) [reply megatron/core/inference/contexts/dynamic_context.py:1]: Addressed

**Review by lmcafee-nvidia — APPROVED** [review]

File: megatron/core/inference/contexts/dynamic_context.py [hunk 3]
```diff
@@ -2875,14 +2875,27 @@ def check_availability(self, req: DynamicInferenceRequest) -> Tuple[bool, bool,
...
-        kv_cache_available = self.kv_block_allocator.is_memory_available(num_blocks_from_pool)
+        # add_request pins the matched blocks before allocating. Only matches that
+        # are currently evictable (ref_count == 0) count against the evictable
+        # pool; matches already pinned by another in-flight request are not in
+        # get_evictable_block_count() and pinning them frees nothing. Reserve only
+        # the ref_count == 0 matches so availability is not under-reported.
+        reserved_evictable = 0
+        if matched_block_ids:
+            matched_tensor = torch.tensor(matched_block_ids, dtype=torch.int32, device='cpu')
+            reserved_evictable = int(
```
Comment (kvareddy) [inline megatron/core/inference/contexts/dynamic_context.py]: name change: potential_matched_count

File: megatron/core/inference/contexts/kv_block_allocator.py [hunk 4]
```diff
@@ -352,14 +425,65 @@ def evict_lru_blocks(self, num_blocks_needed: int) -> bool:
...
+        own_ts = self.block_timestamps[cached_block_ids]
+        hashes = self.block_hashes[cached_block_ids]
+        parents = self.block_parent_hashes[cached_block_ids]
+
+        # Resolve each block's parent hash to the local index of the parent block,
+        # or -1 when the parent is not cached (root block with parent hash 0, or a
+        # parent still in use). Hashes are unique per cached block, so a single
+        # sorted lookup suffices. This O(num_cached log num_cached) setup stays
+        # vectorized; only the inherently-sequential peel below is per-element.
+        sorted_hashes, sort_order = torch.sort(hashes)
```
Comment (kvareddy) [inline megatron/core/inference/contexts/kv_block_allocator.py]: How about maintaining parent_idx?

File: megatron/core/inference/contexts/kv_block_allocator.py [hunk 5]
```diff
@@ -352,14 +425,65 @@ def evict_lru_blocks(self, num_blocks_needed: int) -> bool:
...
+        pos = torch.searchsorted(sorted_hashes, parents).clamp(max=num_cached - 1)
+        found = sorted_hashes[pos] == parents
+        parent_idx = torch.where(found, sort_order[pos], torch.full_like(pos, -1))
+        has_parent = parent_idx >= 0
+
+        # Number of cached children per block; a block is an evictable leaf once
+        # this reaches 0.
+        child_count = torch.zeros(num_cached, dtype=torch.int64)
+        child_count.scatter_add_(
```
Comment (kvareddy) [inline megatron/core/inference/contexts/kv_block_allocator.py]: Same here. Let's maintain the child count.

File: megatron/core/inference/contexts/dynamic_context.py (hunk 3 above)
Comment (santhnm2) [reply megatron/core/inference/contexts/dynamic_context.py]: Done

File: megatron/core/inference/contexts/kv_block_allocator.py (hunk 4 above)
Comment (santhnm2) [reply megatron/core/inference/contexts/kv_block_allocator.py]: Done

File: megatron/core/inference/contexts/kv_block_allocator.py (hunk 5 above)
Comment (santhnm2) [reply megatron/core/inference/contexts/kv_block_allocator.py]: Done

**Review by kvareddy — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]
